### PR TITLE
Add higher-order Sommerfeld boundary conditions for ScalarSelfForce

### PR DESCRIPTION
**Pull request by NamiNishimura73**
## Proposed changes
- Changes related to ScalarSelfForce
- In CircularOrbit, higher order smoothstep has been implemented. Order = 2 case is verified in Test_CircularOrbit 
- Higher order Sommerfeld boundary condition has been implemented. Order = 2 with hyperboloidal slicing case is verified in Test_Sommerfeld 
- lower_edge == upper_edge case has been added in Math.hpp. This case is verified in Test_Math

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Angular.cpp [hunk 1]
```diff
@@ -51,17 +55,35 @@ SPECTRE_TEST_CASE("Unit.ScalarSelfForce.BoundaryConditions.Angular",
...
+    const Direction<2> direction = Direction<2>::lower_xi();
     const DataVector used_for_size(5);
     auto field = make_with_value<Scalar<ComplexDataVector>>(
         used_for_size, std::numeric_limits<double>::signaling_NaN());
     auto n_dot_field_gradient = make_with_value<Scalar<ComplexDataVector>>(
         used_for_size, std::numeric_limits<double>::signaling_NaN());
     const tnsr::i<ComplexDataVector, 2> deriv_field{
         used_for_size.size(), std::numeric_limits<double>::signaling_NaN()};
-    const auto box = db::create<db::AddSimpleTags<>>();
+    const tnsr::I<DataVector, 2> coords{used_for_size.size(), 0.0};
```
Comment (nilsvu) [inline tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Angular.cpp]: You don't need these changes in `Test_Angular.cpp` and `Test_None.cpp`. Please only change `Test_Sommerfeld.cpp`.

File: tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Sommerfeld.cpp [hunk 2]
```diff
@@ -28,55 +31,82 @@
...
+  const std::string base_sommerfeld =
+        "Sommerfeld:\n"
+        "  BlackHoleMass: 1.0\n"
+        "  BlackHoleSpin: 0.5\n"
+        "  OrbitalRadius: 10.0\n"
+        "  MModeNumber: 2\n";
+  const Direction<2> direction = Direction<2>::lower_xi();
+  const size_t num_pts = 5;
   {
-    INFO("Properties");
```
Comment (nilsvu) [inline tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Sommerfeld.cpp:53]: Please don't delete this test of the properties.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Elliptic/Systems/SelfForce/Scalar/AnalyticData/CircularOrbit.cpp [hunk 3]
```diff
@@ -23,15 +23,18 @@
 namespace ScalarSelfForce::AnalyticData {
 
 namespace {
+template <size_t Order>
 std::pair<DataVector, DataVector> boost_function_and_deriv(
     const DataVector& r_star, const std::array<double, 4>& transition_points) {
   return {
-      smoothstep<1>(transition_points[0], transition_points[1], r_star) +
-          smoothstep<1>(transition_points[2], transition_points[3], r_star) -
+      smoothstep<Order>(transition_points[0], transition_points[1], r_star) +
```
Comment (nilsvu) [inline src/Elliptic/Systems/SelfForce/Scalar/AnalyticData/CircularOrbit.cpp]: Please move these changes to the boost function to a separate PR. Keep only the changes to the Sommerfeld boundary conditions in this PR.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Angular.cpp (hunk 1 above)
Comment (NamiNishimura73) [reply tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Angular.cpp]: I have reverted the changes to Test_Angular.cpp and Test_None.cpp to keep them minimal. 
Because now the standard_boundary_conditions requires Beta and Gamma tags, I replaced it with a local tmpl::list specific to the boundary condition being tested.

File: tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Sommerfeld.cpp (hunk 2 above)
Comment (NamiNishimura73) [reply tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Sommerfeld.cpp:53]: It's now included

File: src/Elliptic/Systems/SelfForce/Scalar/AnalyticData/CircularOrbit.cpp (hunk 3 above)
Comment (NamiNishimura73) [reply src/Elliptic/Systems/SelfForce/Scalar/AnalyticData/CircularOrbit.cpp]: removed from this PR

File: tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Sommerfeld.cpp [hunk 4]
```diff
@@ -28,55 +31,94 @@
...
+        "  OrbitalRadius: 10.0\n"
+        "  MModeNumber: 2\n";
+  const Direction<2> direction = Direction<2>::lower_xi();
+  const size_t num_pts = 5;
+
+  {
+    INFO("Apply boundary condition: Order 1");
+    const auto created_1 = TestHelpers::test_factory_creation<BC, Sommerfeld>(
+        base_sommerfeld + "  HyperboloidalSlicing: False\n  Order: 1");
+    const auto& bc_1 = dynamic_cast<const Sommerfeld&>(*created_1);
```
Comment (nilsvu) [inline tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Sommerfeld.cpp]: You don't have to add `_1` labels because these variables are only defined within the scope `{ }` (this is different to Python where scopes are less strict). So you can just keep the names `created` and `boundary_condition` etc here.

**Review by nilsvu — COMMENTED** [review]
Please rebase on the current develop branch, which has the `None` boundary condition already merged. Also please squash your commits into one.

File: tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Sommerfeld.cpp (hunk 4 above)
Comment (NamiNishimura73) [reply tests/Unit/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Test_Sommerfeld.cpp]: Right, thank you!
Rebased on develop, squashed to a single commit, and addressed this.

File: src/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Sommerfeld.hpp [hunk 5]
```diff
@@ -99,20 +108,29 @@ class Sommerfeld : public elliptic::BoundaryConditions::BoundaryCondition<2> {
     return {elliptic::BoundaryConditionType::Neumann};
   }
 
-  using argument_tags = tmpl::list<>;
+  using argument_tags =
+      tmpl::list<domain::Tags::Coordinates<2, Frame::Inertial>, Tags::Beta,
```
Comment (nilsvu) [inline src/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Sommerfeld.hpp]: You can remove the coordinates (not needed)

**Review by nilsvu — COMMENTED** [review]
Looks good! clang-tidy also has some suggestions.

File: src/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Sommerfeld.hpp (hunk 5 above)
Comment (NamiNishimura73) [reply src/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Sommerfeld.hpp]: Clang-tidy fixed and coordinate removed.

File: src/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Sommerfeld.hpp [hunk 6]
```diff
@@ -99,20 +108,27 @@ class Sommerfeld : public elliptic::BoundaryConditions::BoundaryCondition<2> {
...
   void apply(gsl::not_null<Scalar<ComplexDataVector>*> field,
              gsl::not_null<Scalar<ComplexDataVector>*> n_dot_field_gradient,
-             const tnsr::i<ComplexDataVector, 2>& deriv_field) const;
+             const tnsr::i<ComplexDataVector, 2>& deriv_field,
+             const Scalar<ComplexDataVector>& beta,
+             const tnsr::i<ComplexDataVector, 2>& gamma) const;
 
-  using argument_tags_linearized = tmpl::list<>;
+  using argument_tags_linearized =
+      tmpl::list<domain::Tags::Coordinates<2, Frame::Inertial>, Tags::Beta,
```
Comment (nilsvu) [inline src/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Sommerfeld.hpp]: Also remove coordinates here please

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Review by nilsvu — APPROVED** [review]

File: src/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Sommerfeld.hpp (hunk 6 above)
Comment (NamiNishimura73) [reply src/Elliptic/Systems/SelfForce/Scalar/BoundaryConditions/Sommerfeld.hpp]: sorry I missed it! they are removed now.
